core/src/main/java/com/scalar/db/storage/dynamo/DynamoOperation.java

```diff
@@ -64,37 +64,31 @@ public String getGlobalIndexName(String column) {
   @Nonnull
   public Map<String, AttributeValue> getKeyMap() {
     Map<String, AttributeValue> keyMap = new HashMap<>();
-    String partitionKey = getConcatenatedPartitionKey();
-    keyMap.put(PARTITION_KEY, AttributeValue.builder().s(partitionKey).build());
+    ByteBuffer partitionKey = getConcatenatedPartitionKey();
+    keyMap.put(
+        PARTITION_KEY, AttributeValue.builder().b(SdkBytes.fromByteBuffer(partitionKey)).build());
 
     getConcatenatedClusteringKey()
-        .ifPresent(k -> keyMap.put(CLUSTERING_KEY, AttributeValue.builder().s(k).build()));
+        .ifPresent(
+            k ->
+                keyMap.put(
+                    CLUSTERING_KEY,
+                    AttributeValue.builder().b(SdkBytes.fromByteBuffer(k)).build()));
 
     return keyMap;
   }
 
-  String getConcatenatedPartitionKey() {
-    Map<String, Value<?>> keyMap = new HashMap<>();
-    operation.getPartitionKey().get().forEach(v -> keyMap.put(v.getName(), v));
-
-    ConcatenationVisitor visitor = new ConcatenationVisitor();
-    metadata.getPartitionKeyNames().forEach(name -> keyMap.get(name).accept(visitor));
-
-    return visitor.build();
+  ByteBuffer getConcatenatedPartitionKey() {
+    return new KeyBytesEncoder().encode(operation.getPartitionKey());
```

Comment:
We also use `KeyBytesEncoder` to encode the concatenated partition key for efficiency.